src/lib/work_partitioner/work_partitioner.ml

```diff
@@ -1 +1,240 @@
+    Work.With_job_meta.
+      { spec = single_spec
+      ; job_id = pairing
+      ; issued_since_unix_epoch = epoch_now ()
+      ; sok_message
+      }
+  in
+  match single_spec with
+  | Transition (input, witness) -> (
+      match witness.transaction with
+      | Command (Zkapp_command zkapp_command) -> (
+          let witness = Transaction_witness.read_all_proofs_from_disk witness in
+          match
+            Snark_worker_shared.extract_zkapp_segment_works
+              ~m:partitioner.transaction_snark ~input ~witness ~zkapp_command
+          with
+          | Ok (_ :: _ as all) ->
+              let unscheduled_segments =
+                all
+                |> List.map ~f:(fun (witness, spec, statement) ->
+                       Work.Spec.Sub_zkapp.Segment { statement; witness; spec }
+                       |> Work.Spec.Sub_zkapp.read_all_proofs_from_disk )
+                |> Queue.of_list
+              in
+              let pending_zkapp_command =
+                Pending_zkapp_command.create ~job ~unscheduled_segments
+              in
+              let pending_zkapp_command_job =
+                Work.With_job_meta.
+                  { spec = pending_zkapp_command
+                  ; job_id = pairing
+                  ; issued_since_unix_epoch = epoch_now ()
+                  ; sok_message
+                  }
+              in
+              assert (
+                phys_equal `Ok
+                  (Zkapp_command_job_pool.add ~id:pairing
+                     ~job:pending_zkapp_command_job
+                     partitioner.zkapp_command_jobs ) ) ;
+              issue_from_pending_zkapp_command ~partitioner
+                pending_zkapp_command_job
+              |> Option.value_exn
+                   ~message:
+                     "FATAL: we already inserted work into partitioner so this \
+                      shouldn't happen"
+          | Ok [] ->
+              failwith "No witness generated"
+          | Error e ->
+              failwith (Error.to_string_hum e) )
+      | Command (Signed_command _) | Fee_transfer _ | Coinbase _ ->
+          let job =
+            Work.With_job_meta.map
+              ~f_spec:Work.Spec.Single.read_all_proofs_from_disk job
+          in
+          Single { job; data = () } )
+  | Merge _ ->
+      let job =
+        Work.With_job_meta.map
+          ~f_spec:Work.Spec.Single.read_all_proofs_from_disk job
+      in
+      Single { job; data = () }
+
+let issue_from_tmp_slot ~(partitioner : t) () =
+  match partitioner.tmp_slot with
+  | Some spec ->
+      partitioner.tmp_slot <- None ;
+      let single_spec, pairing, sok_message = spec in
+      Some
+        (convert_single_work_from_selector ~partitioner ~single_spec ~pairing
+           ~sok_message )
+  | None ->
+      None
+
+let issue_job_from_partitioner ~(partitioner : t) () :
+    Work.Spec.Partitioned.Stable.Latest.t option =
+  List.find_map
+    ~f:(fun f -> f ())
+    [ reissue_old_zkapp_job ~partitioner
```

Comment:
Could make this a large nested pattern matching, and remove assertion at `Mark woeiqweoqjoqwidh`